Abstract IncentivesContract logic, adapt Staked Incentives Controller and Pull Reward Incentives Controller

kartojal opened the pull request:
### Changes
- Add BaseIncentivesController abstract contract 
- Adapt StakedIncentives to inherit new abstract contract without any internal change.
- Add PullRewardsIncentivesController contract that inherits Base contracts and at claim does ERC20 transfer from external reward vault.
- Add  deployment scripts.

### Reasoning

As @miguelmtzinf pointed out, the last iteration of the BaseIncentivesController had the same logic than the Staked Incentives contracts so makes sense to extract the logic into an abstract contract and them and in this way we can create new customized incentives controllers without touching the core logic  or storage of  the rewards distribution.

kartojal committed:
feat: Added base ERC20 incentives controller contract and deploy scripts.

kartojal committed:
feat: Added tests for base incentives controller

kartojal committed:
fix: fix npm run test script

kartojal commented on contracts/incentives/BaseIncentivesController.sol:
```diff
@@ -0,0 +1,221 @@
+   * @dev returns the current rewards vault contract
+   * @return address
+   */
+  function getRewardsVault() external view returns (address) {
```
I think `getRewardsVault` is not part from the incentives interface due other future implementations doesnt need it. Is not part of the core contract (distribute rewards).

kartojal commented on contracts/incentives/BaseIncentivesController.sol:
```diff
@@ -0,0 +1,221 @@
+
+  uint256 public constant REVISION = 1;
+
+  address public immutable override REWARD_TOKEN;
```
Yes good addition

kartojal commented on contracts/incentives/BaseIncentivesController.sol:
```diff
@@ -0,0 +1,221 @@
+
+/**
+ * @title BaseIncentivesController
+ * @notice Distributor contract for ERC20 rewards to the Aave protocol participants
```
True, will make it more clear

kartojal commented:
> Should we update the `StakedTokenIncentivesController` making it inheriting the `BaseIncentivesController` ?

I think is better to not encapsulate it more deeper.

kartojal commented on test/BaseIncentivesController/claim-on-behalf.spec.ts:
```diff
@@ -0,0 +1,139 @@
+      .to.emit(baseIncentivesController, 'RewardsClaimed')
+      .withArgs(userWithRewards.address, thirdClaimer.address, thirdClaimer.address, '99999');
+    const afterStkBalance = await aaveToken.balanceOf(thirdClaimer.address);
+    console.log('adt', afterStkBalance.toString());
```
True, removing

kartojal committed:
feat: Use an abstract contract to hold all the IncentivesController logic. Adapt Stake and PullRewards contracts to use the abstracted incentives controller.

kartojal committed:
rename: rename incentives proposal to incentives controller in package json and readme.md

kartojal committed:
scripts: fix task name at tests script

kartojal committed:
feat: Added claimRewardsToSelf function to BaseIncentivesController to allow users to claim rewards directly to msg.sender

kartojal committed:
tests: added test case for claimRewardsToSelf

LHerskind commented on contracts/incentives/base/DistributionManager.sol:
```diff
@@ -2,9 +2,9 @@
+import {DistributionTypes} from '../../lib/DistributionTypes.sol';
 
 /**
  * @title DistributionManager
```
- The `AssetData` uses only `104` bits for the `index` but there is still `8` free bits before the word is filled, normally we are using `128` bits for the indexes so think we should at least use the free ones? Also, the `users` could be renamed to `userIndexes` or something like that for clarity.
- Both `getDistributionEnd()` and `DISTRIBUTION_END()` return `_distributionEnd`, are both kept for backwards compatability? 
- `_updateAssetStateInternal` fetches `emissionPerSecond` at every call, but it is not used when `block.timestamp == lastUpdateTimestamp`, could be fetched afterwards. Also the `assetConfig.lastUpdateTimestamp = uint40(block.timestamp);` can be moved out of the if as it is updated in both cases removing the else case.

kartojal replied:
Really like your thoughts, but this PR was aimed  for making IncentivesContract abstract and would not like to touch other core parts like the Distribution Manager.

This can be done in a separate PR.

kartojal replied:
Can you create an issue to track your optimiation?

LHerskind replied:
Added issue #8

LHerskind commented on contracts/incentives/base/BaseIncentivesController.sol:
```diff
@@ -0,0 +1,207 @@
+      assetsConfig[i].underlyingAsset = assets[i];
+      assetsConfig[i].emissionPerSecond = uint104(emissionsPerSecond[i]);
+
+      require(assetsConfig[i].emissionPerSecond == emissionsPerSecond[i], 'INVALID_CONFIGURATION');
```
The overflow check is done differently than similar checks in the `DistributionManager`, could probably be more aligned. 

From `DistributionManager::_updateAssetStateInternal()`:
`require(uint104(newIndex) == newIndex, 'Index overflow');`

kartojal replied:
Done, better to be consistent between contracts.

LHerskind commented on contracts/incentives/base/BaseIncentivesController.sol:
```diff
@@ -0,0 +1,207 @@
+  }
+
+  /// @inheritdoc IAaveIncentivesController
+  function setClaimer(address user, address caller) external override onlyEmissionManager {
```
Should it be possible for a user to set a claimer for themselves? e.g., update `_authorizedClaimers[msg.sender]`, don't see why it needs to be the EmissionManager only, as long as it cannot be set on behalf of other users.

miguelmtzinf replied:
I think this function makes sense as it is since its only callable by the EmissionManager.

kartojal replied:
This function `function setClaimer(address user, address caller)` is intented for 3rd party smart contracts that doesnt support claim from scratch and needs another contract as the incentives distributor for their users. Thats why EmissionManager trusted role is used here.

Would be interesting that end users could also set their own claimers, like a external hot wallet that claims rewards to the cold wallet address, but i would not like to change the current Abstract incentives contract in this PR. Could be added in a separate PR as this one is quite big and only focused into making incentives logic abstract.

kartojal replied:
Can you create an issue to track your optimization? Thanks

LHerskind replied:
Added in #9

LHerskind commented on contracts/incentives/base/BaseIncentivesController.sol:
```diff
@@ -0,0 +1,207 @@
+    if (amount == 0) {
+      return 0;
+    }
+    uint256 unclaimedRewards = _usersUnclaimedRewards[user];
```
If `amount <= unclaimedRewards` it is possible to shortcircuit the claim and just use the unclaimed to payout the reward no need to compute how much have accrued. Expect it to save a lot of gas when the case is true, and only spend little when it is not because we already cached the `unclaimedRewards`.

kartojal replied:
Great find! Will make this change.

kartojal replied:
Done, let me know if it meets your criteria or if you thought another way.

LHerskind reviewed (commented):
Added some comments for the contracts, mostly look good to me. Have not reviewed tests and config yet.

stevenvaleri commented on contracts/incentives/base/BaseIncentivesController.sol:
```diff
@@ -0,0 +1,207 @@
+    address user,
+    uint256 totalSupply,
+    uint256 userBalance
+  ) external override {
```
Does this need to be external?

kartojal replied:
Yes, "handleAction" is called by ATokens or Debt tokens.

stevenvaleri commented on contracts/incentives/PullRewardsIncentivesController.sol:
```diff
@@ -0,0 +1,59 @@
+   * @dev update the rewards vault address, only allowed by the Rewards admin
+   * @param rewardsVault The address of the rewards vault
+   **/
+  function setRewardsVault(address rewardsVault) external onlyEmissionManager {
```
Is there a risk of updating the rewardsVault, but leaving existing token emissions? For token contract updates, would be fine, but if switching tokens could be an issue. Probably can just be handled with clear documentation but can consider include smart contract options.

Potential options
Clearly document that existing emissions will remain for the new token
Clear all emmissions on `setRewardsVault`
Add a parameters to `setRewardsVault` and include a call to update emissions

kartojal replied:
If the EmissionManager changes the rewards vault, it will not change the current token as is inmmutable, just the source where "pull" and transfer the reward token. In my opinion emissions could still be the same and if the "reward vault" has sufficient tokens the incentives will not stop.

The rewards vault can be an EOA or smart contract with the sole purpose of pulling the reward tokens from this contract to the incentivized accounts.

stevenvaleri replied:
Makes sense +1

kartojal commented on contracts/incentives/StakedTokenIncentivesController.sol:
```diff
@@ -19,31 +14,13 @@ import {IAaveIncentivesController} from '../interfaces/IAaveIncentivesController
-  DistributionManager
-{
-  using SafeMath for uint256;
+contract StakedTokenIncentivesController is BaseIncentivesController {
```
True, will remove it

kartojal committed:
feat: perform similar overflow check as DistributionManager

Perform similar overflow check at BaseIncentivesController:58 as DistributionManager::_updateAssetStateInternal()

kartojal committed:
feat: bypass accrue extra rewards at claim if amount is less than unclaimed rewards

kartojal committed:
feat: fix tests, remove unused initialize parameter

kartojal committed:
feat: remove large commented block

LHerskind reviewed (approved):
Looks good to me
